scipy/linalg/_decomp_cossin.py

```diff
@@ -0,0 +1,186 @@
+
+    With dimensions
+    shape(X11) = (p,q)
+    I11: min(p, q) - r
+    I12: min(p, m - q) - r
+    I21: min(m - p, q) - r
+    I22: min(m - p, m - q) - r
+    where r = min(p, m - p, q, m - q)
+
+    Partitioning can be given two ways:
+    1) directly by listing the subblocks
+    2) providing the whole matrix and providing the dimensions of the top left
+        subblock in `partitioning`
+
+    Parameters
+    ----------
+    X : (m,m) array_like or iterable
+        complex unitary or real orthogonal matrix to be decomposed, or iterable
+        of subblocks X11,X12,X21,X22, when ``p=q=None``
+    partitioning : (int, int) or None
+        shape of the upper left block X11, if omitted, X is assumed to be given
+        in iterable form
+
+    Returns
+    -------
+    U : (m, m) ndarray
+        The block diagonal unitary or orthogonal matrix consisting of the
+        blocks U1 (p-by-p) and U2 (m-p)-by-(m-p)
+    CS : (m, m) ndarray
+        The cosine-sine factor with the structure described above
+    VDH : (m, m) ndarray
+        The block diagonal unitary or orthogonal matrix consisting of the
+        blocks V1 (q-by-q) and V2 (m-q)-by-(m-q)
+
+    Examples
+    --------
+    >>> from scipy.linalg import cossin
+    >>> from scipy.stats import unitary_group
+    >>> x = unitary_group.rvs(4)
+    >>> u, cs, vdh = cossin(x, (2,2))
+    >>> np.allclose(x, u @ cs @ vdh)
+    True
+
+    References
+    ----------
+    .. [1] : Brian D. Sutton. Computing the complete CS decomposition. Numer.
+           Algorithms, 50(1):33-65, 2009.
+
+    """
+    if partitioning:
+        X = _asarray_validated(X, check_finite=True)
+        if not np.equal(*X.shape):
+            raise ValueError("Cosine Sine decomposition only supports square"
+                             " matrices, got {}".format(X.shape))
+        p, q = partitioning
+        return _cossin(x11=X[:p, :q], x12=X[:p, q:], x21=X[p:, :q],
+                       x22=X[p:, q:])
+
+    if not isinstance(X, Iterable):
+        raise ValueError("When top_left_shape=None, X must be Iterable")
+
+    X = list(X)
+    if len(X) != 4:
+        raise ValueError("When partitioning=None, "
+                         "exactly four submatrices should be in X, got {}"
+                         .format(len(X)))
+
+    x11, x12, x21, x22 = X
+
+    p, q = np.shape(np.atleast_2d(x11))
+    mmp, mmq = np.shape(np.atleast_2d(x22))
+
+    p2, mmq2 = np.shape(np.atleast_2d(x12))
+    mmp2, q2 = np.shape(np.atleast_2d(x21))
+
+    if p != p2 or q != q2 or mmp != mmp2 or mmq != mmq2 or mmq + q != mmp + p:
+        shapes = [np.shape(np.atleast_2d(subx)) for subx in X]
+        raise ValueError("invalid submatrix dimensions {} to build a square "
+                         "matrix, maybe missing partitioning?".format(shapes))
+    return _cossin(x11, x12, x21, x22)
+
+
+def _cossin(x11, x12, x21, x22):
+    p, q = np.shape(x11)
+    m = np.shape(x22)[0] + p
+
+    iscomp = np.iscomplexobj(x11)
```

Comment:
You have to check all 4 of them to see if anyone is complex.

Reply:
good catch, I was wondering about that while writing...thanks!

Reply:
now this led me down a path of finding the largest type and casting all objects to that. 
Let me know what you think. Also wrote a test to cover this case with mixed types.

Reply:
Nice! However `get_lapack_funs` is designed specifically for this. You just need to provide all the arrays `[x11, x12, x21, x22]` as the second argument and it will give the correct flavor and the rest (up/down casting) is done via `f2py`.

Reply:
Yes, that's how I started - by `np.any(np.iscomplexobj(...)...)` for the prefix decision - but then, I'll still need the largest type explicitly for the CS construction in https://github.com/scipy/scipy/pull/11524/files#diff-aa0b4763c42b00d6594f8e5df1bd90a3R170